packages/hub_domain/lib/src/repositories/post_repository.dart

```diff
@@ -0,0 +1,99 @@
+import 'package:bad_words/bad_words.dart';
+import 'package:hub_domain/src/models/models.dart';
+
+/// Enum for the possible expected failures when creating a post.
+enum CreatePostFailure {
+  /// The post message is too long.
+  tooLong,
+
+  /// The post message includes profanit.
+  isProfane,
+
+  /// Something unexpected happened while creating the post.
+  unexpected,
+}
+
+/// {@template post_creation_failure}
+/// Thrown when a post fails to be created.
+/// {@endtemplate}
+class PostCreationFailure implements Exception {
+  /// {@macro post_creation_failure}
+  const PostCreationFailure(this.reason, this.stackTrace);
+
+  /// The reason why the post failed to be created.
+  final CreatePostFailure reason;
+
+  /// The stack trace associated with the failure.
+  final StackTrace stackTrace;
+}
+
+/// {@template post_repository_adapter}
+/// Adapter for the post repository.
+/// {@endtemplate}
+abstract class PostRepositoryAdapter {
```

Comment:
In the future we might consider using [deposit](https://pub.dev/packages/deposit) here instead. But since this will be the base for the article about code re usage, I didn't wanted to introduce an additional package in the mix. But should definitely be something that we can consider if we want to get more involved with dart full stack.